network/network_interface/network_interface.h

```diff
@@ -0,0 +1,181 @@
+	 * socket.
+	 * @return
+	 *  - \ref SUCCESS : On success
+	 *  - \ref FAILURE : Otherwise
+	 */
+	int32_t (*socket_open)(void *net, uint32_t *sock_id,
+			       enum socket_protocol proto);
+	/**
+	 * @brief Close the socket.
+	 * @param net - Network interface
+	 * @param sock_id - Socket id
+	 * @return
+	 *  - \ref SUCCESS : On success
+	 *  - \ref FAILURE : Otherwise
+	 */
+	int32_t (*socket_close)(void *net, uint32_t sock_id);
+	/**
+	 * @brief Connects TCP socket to a remote host.
+	 * @param net - Network interface
+	 * @param sock_id - Socket id
+	 * @param addr - Remote host
+	 * @return
+	 *  - \ref SUCCESS : On success
+	 *  - \ref FAILURE : Otherwise
+	 */
+	int32_t (*socket_connect)(void *net, uint32_t sock_id,
+				  struct socket_address *addr);
+	/**
+	 * @brief Disconnect form a socket
+	 * @param net - Network interface
+	 * @param sock_id - Socket id
+	 * @return
+	 *  - \ref SUCCESS : On success
+	 *  - \ref FAILURE : Otherwise
+	 */
+	int32_t (*socket_disconnect)(void *net, uint32_t sock_id);
+	/**
+	 * @brief Send data over a TCP socket.
+	 * @param net - Network interface
+	 * @param sock_id - Socket id
+	 * @param data - Buffer of data to send to the host
+	 * @param size - Size of the buffer in bytes
+	 * @return
+	 *  - Number of sent bytes : On success
+	 *  - \ref FAILURE : Otherwise
+	 */
+	int32_t (*socket_send)(void *net, uint32_t sock_id,
+			       const void *data, uint32_t size);
+	/**
+	 * @brief Receive data over a TCP socket.
+	 *
+	 * The socket must be connected to a remote host.
+	 * The call is non blocking
+	 * @param net - Network interface
+	 * @param sock_id - Socket id
+	 * @param data - Destination buffer for received data
+	 * @param size - Maximum data to read
+	 * @return
+	 *  - Number of bytes received into the buffer.
+	 *  - \ref FAILURE is something went wrong
+	 */
+	int32_t (*socket_recv)(void *net, uint32_t sock_id,
+			       void *data, uint32_t size);
+	/**
+	 * @brief Send a packet over a UDP socket.
+	 * @param net - Network interface
+	 * @param sock_id - Socket id
+	 * @param data - Buffer of data to send to the host
+	 * @param size - Size of the buffer in bytes
+	 * @param to - Address of the remote host
+	 * @return
+	 *  - Number of received bytes : On success
+	 *  - \ref FAILURE : Otherwise
+	 */
+	int32_t (*socket_sendto)(void *net, uint32_t sock_id,
+				 const void *data, uint32_t size,
+				 const struct socket_address *to);
+	/**
+	 * @brief Receive a packet over a UDP socket.
+	 * @param net - Network interface
+	 * @param sock_id - Socket id
+	 * @param data - Destination buffer for received data
+	 * @param size - Maximum data to read
+	 * @param from - Destination for the source address or NULL
+	 * @return
+	 *  - Number of sent bytes : On success
+	 *  - \ref FAILURE : Otherwise
+	 */
+	int32_t (*socket_recvfrom)(void *net, uint32_t sock_id,
+				   void *data, uint32_t size,
+				   struct socket_address *from);
```

Comment:
These can be named `socket_sendtcp`, `socket_recvtcp`, `spcket_sendudp` and `socket_recvudp`. This could improve clarity.

Reply:
I tried to keep the same naming as in <sys/socket.h> https://pubs.opengroup.org/onlinepubs/007908799/xns/syssocket.h.html
I am thinking to make them `send `, `recv `, `sendto `and `recvfrom`